src/osd/PrimaryLogPG.cc

```diff
@@ -3466,27 +3467,100 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
+{
+  auto p = refs.begin();
+  while (p!= refs.end()) {
+    dout(10) << __func__ << ": decrement reference on offset oid: " << p->first << dendl;
+    object_locator_t target_oloc(p->first);
+    refcount_manifest(obc, obc->obs.oi.soid, target_oloc, p->first, 
+		      SnapContext(), refcount_t::DECREMENT_REF, NULL);
+    ++p;
+  }
+}
+
+
+void PrimaryLogPG::dec_all_refcount_head_manifest(object_info_t& oi, OpContext* ctx)
+{
+  SnapSetContext* ssc = ctx->obc->ssc;
+  ceph_assert(oi.has_manifest());
+  ceph_assert(oi.soid.is_head());
+  // has snapshot
+  if (ssc && ssc->snapset.clones.size() > 0) {
+    ceph_assert(ssc);
+    dout(15) << __func__ <<  " has snapset " << dendl;
+    object_ref_delta_t refs;
+    ceph_assert(oi.manifest.is_chunked());
+    ceph_assert(!oi.manifest.is_redirect());
+
+    hobject_t clone_oid = oi.soid;
+    clone_oid.snap = ssc->snapset.clone_overlap.rbegin()->first;
+    ObjectContextRef cobc = get_object_context(clone_oid, false, NULL);
+    if (!cobc) {
+      dout(0) << __func__ << ": Can not find clone obc " << clone_oid << dendl;;
+      return;
+    }
+    object_info_t& coi = cobc->obs.oi;
+    oi.manifest.calc_refs_to_drop_on_removal(nullptr, &coi.manifest, refs);
+
+    if (!refs.is_empty()) {
+      ctx->register_on_commit(
+	[oi, ctx, this, refs](){
+	  dec_refcount_non_intersection(ctx->obc, oi, refs);
+      });
+    }
+    return;
+  }
+
+  // no snapshot
+  if ((oi.flags & object_info_t::FLAG_REDIRECT_HAS_REFERENCE) && oi.manifest.is_redirect()) {
+    ctx->register_on_commit(
+      [oi, ctx, this](){
+      object_locator_t target_oloc(oi.manifest.redirect_target);
+      refcount_manifest(ctx->obc, oi.soid, target_oloc, oi.manifest.redirect_target, 
+			SnapContext(), refcount_t::DECREMENT_REF, NULL);
+    });
+  } else if (oi.manifest.is_chunked()) {
+    ctx->register_on_commit(
+      [oi, ctx, this](){
+      for (auto p : oi.manifest.chunk_map) {
+	if (p.second.has_reference()) {
+	  object_locator_t target_oloc(p.second.oid);
+	  refcount_manifest(ctx->obc, oi.soid, target_oloc, p.second.oid, 
+			    SnapContext(), refcount_t::DECREMENT_REF, NULL);
+	}
+      }
+    });
+  } else {
+    ceph_abort_msg("unrecognized manifest type");
+  }
+}
+
+void PrimaryLogPG::refcount_manifest(ObjectContextRef obc, hobject_t src_soid, object_locator_t oloc,
+				     hobject_t tgt_soid, SnapContext snapc, refcount_t type, RefCountCallback* cb)
 {
   unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY |
                    CEPH_OSD_FLAG_RWORDERED;                      
 
-  dout(10) << __func__ << " Start refcount for " << soid << dendl;
+  dout(10) << __func__ << " Start refcount from " << src_soid 
+	   << " to " << tgt_soid << dendl;
     
   ObjectOperation obj_op;
   bufferlist in;
-  if (get) {             
+  if (type == refcount_t::INCREMENT_REF) {             
     cls_chunk_refcount_get_op call;
-    call.source = obc->obs.oi.soid;
+    call.source = src_soid;
```

Comment:
I didn't notice this before, but have you considered what would happen if head references a cas object, we clone the object, and then overwrite that portion of head?  We wouldn't have performed a second ref operation, but the cas object would still have head as the source though head no longer has a reference.  I think the answer is to *always* set head as the owner regardless of the clone, so here I think you want src_soid.head().